lib/rucio/core/rule.py

```diff
@@ -427,33 +474,57 @@ def add_rules(dids, rules, *, session: "Session", logger=logging.log):
             # 2. Get the did
             with METRICS.timer('add_rules.get_did'):
                 try:
-                    did = session.query(models.DataIdentifier).filter(
-                        models.DataIdentifier.scope == elem['scope'],
-                        models.DataIdentifier.name == elem['name']).one()
+                    stmt = select(
+                        models.DataIdentifier
+                    ).where(
+                        and_(models.DataIdentifier.scope == elem['scope'],
+                             models.DataIdentifier.name == elem['name'])
+                    )
+                    did = session.execute(stmt).scalar_one()
                 except NoResultFound as exc:
                     raise DataIdentifierNotFound('Data identifier %s:%s is not valid.' % (elem['scope'], elem['name'])) from exc
                 except TypeError as error:
                     raise InvalidObject(error.args) from error
 
             # 2.1 If the did is a constituent, relay the rule to the archive
             if did.did_type == DIDType.FILE and did.constituent:  # Check if a single replica of this DID exists
-                replica_cnt = session.query(models.RSEFileAssociation).join(models.RSE, models.RSEFileAssociation.rse_id == models.RSE.id)\
-                    .filter(models.RSEFileAssociation.scope == did.scope,
-                            models.RSEFileAssociation.name == did.name,
-                            models.RSEFileAssociation.state == ReplicaState.AVAILABLE,
-                            models.RSE.rse_type != RSEType.TAPE).count()
+                stmt = select(
+                    func.count(models.RSEFileAssociation.rse_id)
```

Comment:
This could be rewritten in the same way as the other queries outlined here https://github.com/rucio/rucio/pull/6493#discussion_r1517310447

Reply:
Good catch, changed this